src/transformers/models/canine/modeling_canine.py

```diff
@@ -0,0 +1,1744 @@
+        # In this transformation, we change the dimensionality from `char_dim` to
+        # `molecule_dim`, but do *NOT* add a resnet connection. Instead, we rely on
+        # the resnet connections (a) from the final char transformer stack back into
+        # the original char transformer stack and (b) the resnet connections from
+        # the final char transformer stack back into the deep BERT stack of
+        # molecules.
+        #
+        # Empirically, it is critical to use a powerful enough transformation here:
+        # mean pooling causes training to diverge with huge gradient norms in this
+        # region of the model; using a convolution here resolves this issue. From
+        # this, it seems that molecules and characters require a very different
+        # feature space; intuitively, this makes sense.
+        init_molecule_encoding = self.chars_to_molecules(input_char_encoding)
+
+        # Deep BERT encoder
+        # `molecule_sequence_output`: shape (batch_size, mol_seq_len, mol_dim)
+        encoder_outputs = self.encoder(
+            init_molecule_encoding,
+            attention_mask=extended_molecule_attention_mask,
+            head_mask=head_mask,
+            encoder_hidden_states=encoder_hidden_states,
+            encoder_attention_mask=encoder_extended_attention_mask,
+            past_key_values=past_key_values,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+        molecule_sequence_output = encoder_outputs[0]
+        pooled_output = self.pooler(molecule_sequence_output) if self.pooler is not None else None
+
+        # Upsample molecules back to characters.
+        # `repeated_molecules`: shape (batch_size, char_seq_len, mol_hidden_size)
+        repeated_molecules = self._repeat_molecules(molecule_sequence_output, char_seq_length=input_shape[-1])
+
+        # Concatenate representations (contextualized char embeddings and repeated molecules):
+        # `concat`: shape [batch_size, char_seq_len, molecule_hidden_size+char_hidden_final]
+        concat = torch.cat([input_char_encoding, repeated_molecules], dim=-1)
+
+        # Project representation dimension back to hidden_size
+        # `sequence_output`: shape (batch_size, char_seq_len, hidden_size])
+        sequence_output = self.projection(concat)
+
+        # Apply final shallow Transformer
+        # `sequence_output`: shape (batch_size, char_seq_len, hidden_size])
+        final_chars_encoder_outputs = self.final_char_encoder(
+            sequence_output,
+            extended_attention_mask,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+        )
+        sequence_output = final_chars_encoder_outputs.last_hidden_state
+
+        if not return_dict:
+            return (sequence_output, pooled_output) + final_chars_encoder_outputs[1:]
```

Comment:
Hi @NielsRogge , with one of the recent commits, it is no longer possible to get hidden states from all layers.

Normally I would use:

```
model_name = "nielsr/canine-s"

config = AutoConfig.from_pretrained(model_name, output_hidden_states=True)
model = CanineModel.from_pretrained(pretrained_model_name_or_path=model_name, config=config)
tokenizer = CanineTokenizer.from_pretrained(model_name)
```

then:

```
encoding = tokenizer([sent.to_tokenized_string()], padding="longest", return_tensors="pt")
hidden_states = model(**encoding).hidden_states
print(len(hidden_states))
```

should output all layers (instead of 2 here)

Reply:
Yes, but as mentioned in the first post of this PR, I'm not sure whether CANINE should output a regular `ModelOutput` as other models, since it actually uses 3 Transformer encoders. So perhaps I'll define a custom `CanineModelOutput`.